// @trellisorg/ngrx-data-websocket-core/src/lib/types/socket-op.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.listeners = exports.dispatchEvents = exports.reservedEvents = exports.SocketOp = exports.ROOT_ACTION_VALUE = void 0;
exports.ROOT_ACTION_VALUE = 'ngrx-data-websocket';
var SocketOp;
(function (SocketOp) {
    SocketOp["CONNECT"] = "connect";
    SocketOp["CONNECT_ERROR"] = "connect_error";
    SocketOp["CONNECT_TIMEOUT"] = "connect_timeout";
    SocketOp["RECONNECT"] = "reconnect";
    SocketOp["RECONNECT_ATTEMPT"] = "reconnect_attempt";
    SocketOp["RECONNECTING"] = "reconnecting";
    SocketOp["RECONNECT_ERROR"] = "reconnect_error";
    SocketOp["RECONNECT_FAILED"] = "reconnect_failed";
    SocketOp["SAVE_ADD_MANY"] = "ngrx-data-websocket/save/add-many";
    SocketOp["SAVE_ADD_MANY_ERROR"] = "ngrx-data-websocket/save/add-many/error";
    SocketOp["SAVE_ADD_MANY_SUCCESS"] = "ngrx-data-websocket/save/add-many/success";
    SocketOp["SAVE_ADD_ONE"] = "ngrx-data-websocket/save/add-one";
    SocketOp["SAVE_ADD_ONE_ERROR"] = "ngrx-data-websocket/save/add-one/error";
    SocketOp["SAVE_ADD_ONE_SUCCESS"] = "ngrx-data-websocket/save/add-one/success";
    SocketOp["SAVE_DELETE_ONE"] = "ngrx-data-websocket/save/delete-one";
    SocketOp["SAVE_DELETE_ONE_SUCCESS"] = "ngrx-data-websocket/save/delete-one/success";
    SocketOp["SAVE_DELETE_ONE_ERROR"] = "ngrx-data-websocket/save/delete-one/error";
    SocketOp["QUERY_ALL"] = "ngrx-data-websocket/query-all";
    SocketOp["QUERY_ALL_SUCCESS"] = "ngrx-data-websocket/query-all/success";
    SocketOp["QUERY_ALL_ERROR"] = "ngrx-data-websocket/query-all/error";
    SocketOp["QUERY_BY_KEY"] = "ngrx-data-websocket/query-by-key";
    SocketOp["QUERY_BY_KEY_SUCCESS"] = "ngrx-data-websocket/query-by-key/success";
    SocketOp["QUERY_BY_KEY_ERROR"] = "ngrx-data-websocket/query-by-key/error";
    SocketOp["QUERY_MANY"] = "ngrx-data-websocket/query-many";
    SocketOp["QUERY_MANY_SUCCESS"] = "ngrx-data-websocket/query-many/success";
    SocketOp["QUERY_MANY_ERROR"] = "ngrx-data-websocket/query-many/error";
    SocketOp["SAVE_UPDATE_ONE"] = "ngrx-data-websocket/save/update-one";
    SocketOp["SAVE_UPDATE_ONE_SUCCESS"] = "ngrx-data-websocket/save/update-one/success";
    SocketOp["SAVE_UPDATE_ONE_ERROR"] = "ngrx-data-websocket/save/update-one/error";
    SocketOp["SAVE_UPSERT_ONE"] = "ngrx-data-websocket/save/upsert-one";
    SocketOp["SAVE_UPSERT_ONE_SUCCESS"] = "ngrx-data-websocket/save/upsert-one/success";
    SocketOp["SAVE_UPSERT_ONE_ERROR"] = "ngrx-data-websocket/save/upsert-one/error";
})(SocketOp = exports.SocketOp || (exports.SocketOp = {}));
exports.reservedEvents = [
    SocketOp.CONNECT,
    SocketOp.CONNECT_ERROR,
    SocketOp.CONNECT_TIMEOUT,
    SocketOp.RECONNECT,
    SocketOp.RECONNECT_ATTEMPT,
    SocketOp.RECONNECTING,
    SocketOp.RECONNECT_ERROR,
    SocketOp.RECONNECT_FAILED,
];
exports.dispatchEvents = [
    SocketOp.SAVE_ADD_ONE,
    SocketOp.SAVE_DELETE_ONE,
    SocketOp.QUERY_ALL,
    SocketOp.QUERY_BY_KEY,
    SocketOp.QUERY_MANY,
    SocketOp.SAVE_UPDATE_ONE,
    SocketOp.SAVE_UPSERT_ONE,
];
exports.listeners = [
    SocketOp.SAVE_ADD_MANY_SUCCESS,
    SocketOp.SAVE_ADD_MANY_ERROR,
    SocketOp.SAVE_ADD_ONE_SUCCESS,
    SocketOp.SAVE_ADD_ONE_ERROR,
    SocketOp.SAVE_DELETE_ONE_SUCCESS,
    SocketOp.SAVE_DELETE_ONE_ERROR,
    SocketOp.QUERY_ALL_SUCCESS,
    SocketOp.QUERY_ALL_ERROR,
    SocketOp.QUERY_BY_KEY_SUCCESS,
    SocketOp.QUERY_BY_KEY_ERROR,
    SocketOp.QUERY_MANY_SUCCESS,
    SocketOp.QUERY_MANY_ERROR,
    SocketOp.SAVE_UPDATE_ONE_SUCCESS,
    SocketOp.SAVE_UPDATE_ONE_ERROR,
    SocketOp.SAVE_UPSERT_ONE_SUCCESS,
    SocketOp.SAVE_UPSERT_ONE_ERROR,
];
