// @trellisorg/ngrx-data-websocket-client/esm2020/lib/dispatchers/socket-dispatcher-factory.mjs
import { Inject, Injectable } from '@angular/core';
import { CorrelationIdGenerator } from '@ngrx/data';
import { ScannedActionsSubject, Store } from '@ngrx/store';
import { Observable } from 'rxjs';
import { shareReplay } from 'rxjs/operators';
import { SocketActionFactory } from '../actions/socket-action-factory';
import { defaultNgrxDataWebsocketConfig, NGRX_DATA_WEBSOCKET_CONFIG, } from '../utils/tokens';
import { SocketDispatcherBase } from './socket-dispatcher-base';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/data";
import * as i2 from "../actions/socket-action-factory";
import * as i3 from "@ngrx/store";
import * as i4 from "rxjs";
export class SocketDispatcherFactory {
    constructor(scannedActions$, config, correlationIdGenerator, socketActionFactory, store) {
        this.config = config;
        this.correlationIdGenerator = correlationIdGenerator;
        this.socketActionFactory = socketActionFactory;
        this.store = store;
        this.reducedActions$ = scannedActions$.pipe(shareReplay(1));
        this.raSubscription = this.reducedActions$.subscribe();
    }
    create(entityName) {
        return new SocketDispatcherBase(entityName, this.correlationIdGenerator, this.socketActionFactory, this.reducedActions$, this.store, isNaN(this.config.timeout)
            ? defaultNgrxDataWebsocketConfig.timeout
            : this.config.timeout);
    }
    ngOnDestroy() {
        this.raSubscription.unsubscribe();
    }
}
SocketDispatcherFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.0", ngImport: i0, type: SocketDispatcherFactory, deps: [{ token: ScannedActionsSubject }, { token: NGRX_DATA_WEBSOCKET_CONFIG }, { token: i1.CorrelationIdGenerator }, { token: i2.SocketActionFactory }, { token: i3.Store }], target: i0.ɵɵFactoryTarget.Injectable });
SocketDispatcherFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.0.0", ngImport: i0, type: SocketDispatcherFactory });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.0", ngImport: i0, type: SocketDispatcherFactory, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i4.Observable, decorators: [{
                    type: Inject,
                    args: [ScannedActionsSubject]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [NGRX_DATA_WEBSOCKET_CONFIG]
                }] }, { type: i1.CorrelationIdGenerator }, { type: i2.SocketActionFactory }, { type: i3.Store }]; } });
