// @trellisorg/ngrx-data-websocket-client/esm2020/lib/ngrx-data-websocket-client.module.mjs
import { NgModule } from '@angular/core';
import { EffectsModule, EffectSources } from '@ngrx/effects';
import { SocketActionFactory } from './actions/socket-action-factory';
import { SocketDataServiceFactory } from './data-services/socket-data-service-factory';
import { SocketDispatcherFactory } from './dispatchers/socket-dispatcher-factory';
import { SocketDispatcherEffects } from './effects/socket-dispatcher-effects.service';
import { SocketEventListenerCollectionService } from './listeners/socket-event-listener-collection.service';
import { SocketEventListenerFactory } from './listeners/socket-event-listener-factory';
import { SocketSelectors$Factory } from './selectors/socket-selectors$';
import { SocketServiceElementsFactory } from './socket-services/socket-service-elements.factory';
import { defaultNgrxDataWebsocketConfig, NGRX_DATA_WEBSOCKET_CONFIG, } from './utils/tokens';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/effects";
import * as i2 from "./effects/socket-dispatcher-effects.service";
export class NgrxDataWebsocketClientModule {
    constructor(effectsSources, socketDispatcherEffects) {
        this.effectsSources = effectsSources;
        this.socketDispatcherEffects = socketDispatcherEffects;
        this.effectsSources.addEffects(socketDispatcherEffects);
    }
    static forRoot(config) {
        return {
            ngModule: NgrxDataWebsocketClientModule,
            providers: [
                {
                    provide: NGRX_DATA_WEBSOCKET_CONFIG,
                    useValue: {
                        ...defaultNgrxDataWebsocketConfig,
                        ...config,
                    },
                },
            ],
        };
    }
}
NgrxDataWebsocketClientModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.0", ngImport: i0, type: NgrxDataWebsocketClientModule, deps: [{ token: i1.EffectSources }, { token: i2.SocketDispatcherEffects }], target: i0.ɵɵFactoryTarget.NgModule });
NgrxDataWebsocketClientModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.0.0", ngImport: i0, type: NgrxDataWebsocketClientModule, imports: [EffectsModule] });
NgrxDataWebsocketClientModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.0.0", ngImport: i0, type: NgrxDataWebsocketClientModule, providers: [
        SocketServiceElementsFactory,
        SocketDispatcherFactory,
        SocketActionFactory,
        SocketDataServiceFactory,
        SocketEventListenerFactory,
        SocketEventListenerCollectionService,
        SocketDispatcherEffects,
        SocketSelectors$Factory,
    ], imports: [EffectsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.0", ngImport: i0, type: NgrxDataWebsocketClientModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [EffectsModule],
                    providers: [
                        SocketServiceElementsFactory,
                        SocketDispatcherFactory,
                        SocketActionFactory,
                        SocketDataServiceFactory,
                        SocketEventListenerFactory,
                        SocketEventListenerCollectionService,
                        SocketDispatcherEffects,
                        SocketSelectors$Factory,
                    ],
                }]
        }], ctorParameters: function () { return [{ type: i1.EffectSources }, { type: i2.SocketDispatcherEffects }]; } });
