// @trellisorg/ngrx-data-websocket-client/esm2020/lib/socket-services/socket-service-elements.factory.mjs
import { Inject, Injectable } from '@angular/core';
import { EntityDataService } from '@ngrx/data';
import { SocketDataServiceFactory } from '../data-services/socket-data-service-factory';
import { SocketDispatcherFactory } from '../dispatchers/socket-dispatcher-factory';
import { SocketEventListenerCollectionService } from '../listeners/socket-event-listener-collection.service';
import { SocketEventListenerFactory } from '../listeners/socket-event-listener-factory';
import { SocketSelectors$Factory, } from '../selectors/socket-selectors$';
import { NGRX_DATA_WEBSOCKET_CONFIG, } from '../utils/tokens';
import * as i0 from "@angular/core";
import * as i1 from "../dispatchers/socket-dispatcher-factory";
import * as i2 from "../listeners/socket-event-listener-factory";
import * as i3 from "../data-services/socket-data-service-factory";
import * as i4 from "@ngrx/data";
import * as i5 from "../listeners/socket-event-listener-collection.service";
import * as i6 from "../selectors/socket-selectors$";
export class SocketServiceElementsFactory {
    constructor(config, socketDispatcherFactory, socketEventListenerFactory, socketDataServiceFactory, entityDataService, socketEventListenerCollectionService, socketSelectors$Factory) {
        this.config = config;
        this.socketDispatcherFactory = socketDispatcherFactory;
        this.socketEventListenerFactory = socketEventListenerFactory;
        this.socketDataServiceFactory = socketDataServiceFactory;
        this.entityDataService = entityDataService;
        this.socketEventListenerCollectionService = socketEventListenerCollectionService;
        this.socketSelectors$Factory = socketSelectors$Factory;
    }
    create(entityName) {
        const dispatcher = this.socketDispatcherFactory.create(entityName);
        const listener = this.socketEventListenerFactory.create(entityName);
        const selectors$ = this.socketSelectors$Factory.create(entityName);
        this.socketEventListenerCollectionService.register(entityName, listener);
        const dataService = this.socketDataServiceFactory.create(entityName, dispatcher);
        this.entityDataService.registerService(entityName, dataService);
        return {
            listener,
            dispatcher,
            entityName,
            dataService,
            selectors$,
        };
    }
}
SocketServiceElementsFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.0", ngImport: i0, type: SocketServiceElementsFactory, deps: [{ token: NGRX_DATA_WEBSOCKET_CONFIG }, { token: i1.SocketDispatcherFactory }, { token: i2.SocketEventListenerFactory }, { token: i3.SocketDataServiceFactory }, { token: i4.EntityDataService }, { token: i5.SocketEventListenerCollectionService }, { token: i6.SocketSelectors$Factory }], target: i0.ɵɵFactoryTarget.Injectable });
SocketServiceElementsFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.0.0", ngImport: i0, type: SocketServiceElementsFactory });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.0", ngImport: i0, type: SocketServiceElementsFactory, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [NGRX_DATA_WEBSOCKET_CONFIG]
                }] }, { type: i1.SocketDispatcherFactory }, { type: i2.SocketEventListenerFactory }, { type: i3.SocketDataServiceFactory }, { type: i4.EntityDataService }, { type: i5.SocketEventListenerCollectionService }, { type: i6.SocketSelectors$Factory }]; } });
