// @trellisorg/ngrx-data-websocket-client/esm2020/lib/effects/socket-dispatcher-effects.service.mjs
import { Injectable } from '@angular/core';
import { Actions, createEffect } from '@ngrx/effects';
import { dispatchEvents, listeners, } from '@trellisorg/ngrx-data-websocket-core';
import { filter, map } from 'rxjs/operators';
import { SocketActionFactory } from '../actions/socket-action-factory';
import { SocketEventListenerCollectionService } from '../listeners/socket-event-listener-collection.service';
import { ofSocketType } from '../utils/of_socket_type';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/effects";
import * as i2 from "../listeners/socket-event-listener-collection.service";
import * as i3 from "../actions/socket-action-factory";
export class SocketDispatcherEffects {
    constructor(_actions$, socketEventListenerCollectionService, socketActionFactory) {
        this._actions$ = _actions$;
        this.socketEventListenerCollectionService = socketEventListenerCollectionService;
        this.socketActionFactory = socketActionFactory;
        this.socketEventWithoutCrid$ = createEffect(() => this._actions$.pipe(ofSocketType(...listeners), filter((action) => !action.payload.correlationId), map((action) => this.socketActionFactory.convertToDataAction(action))));
        this.dispatchSocketEvent$ = createEffect(() => this._actions$.pipe(ofSocketType(...dispatchEvents), map((action) => action.payload), map((action) => {
            const listener = this.socketEventListenerCollectionService.get(action.entityName);
            if (listener) {
                listener.emit(action.socketOp, action.correlationId, action.data);
            }
            return;
        })), { dispatch: false });
    }
}
SocketDispatcherEffects.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.0", ngImport: i0, type: SocketDispatcherEffects, deps: [{ token: i1.Actions }, { token: i2.SocketEventListenerCollectionService }, { token: i3.SocketActionFactory }], target: i0.ɵɵFactoryTarget.Injectable });
SocketDispatcherEffects.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.0.0", ngImport: i0, type: SocketDispatcherEffects });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.0", ngImport: i0, type: SocketDispatcherEffects, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.Actions }, { type: i2.SocketEventListenerCollectionService }, { type: i3.SocketActionFactory }]; } });
