// @trellisorg/ngrx-data-websocket-client/esm2020/lib/actions/socket-action-factory.mjs
import { Injectable } from '@angular/core';
import { EntityActionFactory } from '@ngrx/data';
import { createActionType } from '../utils/create-action-type';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/data";
export class SocketActionFactory {
    constructor(entityActionFactory) {
        this.entityActionFactory = entityActionFactory;
    }
    create(entityName, socketOp, data, options) {
        return {
            type: createActionType(entityName, socketOp),
            payload: {
                entityName,
                socketOp,
                data,
                ...options,
            },
        };
    }
    convertToDataAction(socketAction) {
        return this.entityActionFactory.create(socketAction.payload.entityName, socketAction.payload.socketOp.replace('ngrx-data-websocket', '@ngrx/data'), socketAction.payload.data);
    }
}
SocketActionFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.0", ngImport: i0, type: SocketActionFactory, deps: [{ token: i1.EntityActionFactory }], target: i0.ɵɵFactoryTarget.Injectable });
SocketActionFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.0.0", ngImport: i0, type: SocketActionFactory });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.0", ngImport: i0, type: SocketActionFactory, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.EntityActionFactory }]; } });
